### ISSUE TITLE
OOM error while compiling

### ISSUE THREAD

**Issue by cridmann** [scodec/scodec#73]
Hello,

I'm currently using `v1.7.1`

I have a small/moderately sized scodec project, and it appears I've reached a tipping point with the compiler somewhere.  It will take over 10 minutes to compile, and will then throw an out of memory error.  I have 2GB allocated to sbt.

Here is a stack trace:

```
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: GC overhead limit exceeded
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.util.concurrent.FutureTask.get(FutureTask.java:192)
    at sbt.ConcurrentRestrictions$$anon$4.take(ConcurrentRestrictions.scala:188)
    at sbt.Execute.next$1(Execute.scala:83)
    at sbt.Execute.processAll(Execute.scala:86)
    at sbt.Execute.runKeep(Execute.scala:66)
    at sbt.EvaluateTask$.liftedTree1$1(EvaluateTask.scala:342)
    at sbt.EvaluateTask$.run$1(EvaluateTask.scala:341)
    at sbt.EvaluateTask$.runTask(EvaluateTask.scala:361)
    at sbt.Aggregation$$anonfun$3.apply(Aggregation.scala:64)
    at sbt.Aggregation$$anonfun$3.apply(Aggregation.scala:62)
    at sbt.EvaluateTask$.withStreams(EvaluateTask.scala:293)
    at sbt.Aggregation$.timedRun(Aggregation.scala:62)
    at sbt.Aggregation$.runTasks(Aggregation.scala:71)
    at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:32)
    at sbt.Aggregation$$anonfun$applyTasks$1.apply(Aggregation.scala:31)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:60)
    at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:60)
    at sbt.Aggregation$$anonfun$evaluatingParser$4$$anonfun$apply$5.apply(Aggregation.scala:153)
    at sbt.Aggregation$$anonfun$evaluatingParser$4$$anonfun$apply$5.apply(Aggregation.scala:152)
    at sbt.Act$$anonfun$sbt$Act$$actParser0$1$$anonfun$sbt$Act$$anonfun$$evaluate$1$1$$anonfun$apply$10.apply(Act.scala:244)
    at sbt.Act$$anonfun$sbt$Act$$actParser0$1$$anonfun$sbt$Act$$anonfun$$evaluate$1$1$$anonfun$apply$10.apply(Act.scala:241)
    at sbt.Command$.process(Command.scala:92)
    at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
    at sbt.MainLoop$$anonfun$1$$anonfun$apply$1.apply(MainLoop.scala:98)
    at sbt.State$$anon$1.process(State.scala:184)
    at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
    at sbt.MainLoop$$anonfun$1.apply(MainLoop.scala:98)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.MainLoop$.next(MainLoop.scala:98)
    at sbt.MainLoop$.run(MainLoop.scala:91)
    at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:70)
    at sbt.MainLoop$$anonfun$runWithNewLog$1.apply(MainLoop.scala:65)
    at sbt.Using.apply(Using.scala:24)
    at sbt.MainLoop$.runWithNewLog(MainLoop.scala:65)
    at sbt.MainLoop$.runAndClearLast(MainLoop.scala:48)
    at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:32)
    at sbt.MainLoop$.runLogged(MainLoop.scala:24)
    at sbt.StandardMain$.runManaged(Main.scala:53)
    at sbt.xMain.run(Main.scala:28)
    at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:57)
    at xsbt.boot.Launch$.withContextLoader(Launch.scala:77)
    at xsbt.boot.Launch$.run(Launch.scala:57)
    at xsbt.boot.Launch$$anonfun$explicit$1.apply(Launch.scala:45)
    at xsbt.boot.Launch$.launch(Launch.scala:65)
    at xsbt.boot.Launch$.apply(Launch.scala:16)
    at xsbt.boot.Boot$.runImpl(Boot.scala:32)
    at xsbt.boot.Boot$.main(Boot.scala:21)
    at xsbt.boot.Boot.main(Boot.scala)
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
    at scala.collection.mutable.ListBuffer.$plus$eq(ListBuffer.scala:170)
    at scala.collection.immutable.List.loop$1(List.scala:184)
    at scala.collection.immutable.List.mapConserve(List.scala:189)
    at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:115)
    at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:716)
    at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:685)
    at scala.collection.immutable.List.loop$1(List.scala:173)
    at scala.collection.immutable.List.mapConserve(List.scala:189)
    at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:157)
    at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:716)
    at scala.reflect.internal.Types$Type.subst(Types.scala:705)
    at scala.reflect.internal.Types$Type.instantiateTypeParams(Types.scala:470)
    at scala.reflect.internal.Types$ArgsTypeRef.asSeenFromInstantiated$1(Types.scala:1855)
    at scala.reflect.internal.Types$ArgsTypeRef.transform(Types.scala:1882)
    at scala.reflect.internal.Types$AliasTypeRef$class.betaReduce(Types.scala:2043)
    at scala.reflect.internal.Types$AliasArgsTypeRef.betaReduce(Types.scala:2326)
    at scala.reflect.internal.Types$AliasTypeRef$class.dealias(Types.scala:2014)
    at scala.reflect.internal.Types$AliasArgsTypeRef.dealias(Types.scala:2326)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.loop$1(Implicits.scala:519)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.checkCompatibility(Implicits.scala:556)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.isPlausiblyCompatible(Implicits.scala:369)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation.survives(Implicits.scala:816)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19$$anonfun$20.apply(Implicits.scala:872)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19$$anonfun$20.apply(Implicits.scala:872)
    at scala.collection.TraversableLike$$anonfun$filterImpl$1.apply(TraversableLike.scala:259)
    at scala.collection.immutable.List.foreach(List.scala:381)
    at scala.collection.TraversableLike$class.filterImpl(TraversableLike.scala:258)
    at scala.collection.TraversableLike$class.filter(TraversableLike.scala:270)
    at scala.collection.AbstractTraversable.filter(Traversable.scala:104)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19.apply(Implicits.scala:872)
    at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation$$anonfun$19.apply(Implicits.scala:871)
    at scala.collection.immutable.List.flatMap(List.scala:327)
[error] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: GC overhead limit exceeded
```

Some of my case classes are large (some with 38 fields), I use `dropUnits` and `Discriminator` frequently.  I also use scala `Enumeration`s heavily.  Just from the stack trace it seems lost searching for implicits or something.  I can try monkeying around with how I'm using implicits, as I'm actually not relying on that feature too heavily.

It's worth noting that my compile times were pretty quick until recently when I added a few case classes that had many fields in them.  I see https://github.com/scodec/scodec/issues/41 was fixed in `1.7.0` so I doubt the same thing is going on.

Let me know if you'd like some more information or maybe a strip-downed gist of my coding style with scodec.

Thanks!

**Comment by mpilquist** [comment]
Ouch!

If you can, try with scodec-core 1.8.1, just to rule out any updates in Shapeless that might fix this. Assuming that doesn't work, a gist that exhibits the problem would be much appreciated.

Cheers

**Comment by cridmann** [comment]
With `1.8.1`, it still failed though with a bit different of a trace:

```
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
	at scala.reflect.internal.Scopes$class.newScope(Scopes.scala:399)
	at scala.reflect.internal.SymbolTable.newScope(SymbolTable.scala:16)
	at scala.reflect.internal.Types$class.refinedType(Types.scala:3432)
	at scala.reflect.internal.SymbolTable.refinedType(SymbolTable.scala:16)
	at scala.reflect.internal.Types$class.copyRefinedType(Types.scala:3438)
	at scala.reflect.internal.SymbolTable.copyRefinedType(SymbolTable.scala:16)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:159)
	at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:716)
	at scala.reflect.internal.Types$Type.subst(Types.scala:705)
	at scala.reflect.internal.Types$Type.instantiateTypeParams(Types.scala:470)
	at scala.reflect.internal.Types$ArgsTypeRef.asSeenFromInstantiated$1(Types.scala:1855)
	at scala.reflect.internal.Types$ArgsTypeRef.transform(Types.scala:1882)
	at scala.reflect.internal.Types$AliasTypeRef$class.betaReduce(Types.scala:2043)
	at scala.reflect.internal.Types$AliasArgsTypeRef.betaReduce(Types.scala:2326)
	at scala.reflect.internal.Types$AliasTypeRef$class.dealias(Types.scala:2014)
	at scala.reflect.internal.Types$AliasArgsTypeRef.dealias(Types.scala:2326)
	at scala.reflect.internal.Types$Type.dealiasWiden(Types.scala:526)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.scala$tools$nsc$typechecker$Implicits$ImplicitSearch$$core$1(Implicits.scala:387)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.scala$tools$nsc$typechecker$Implicits$ImplicitSearch$$dominates(Implicits.scala:414)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$$anonfun$14.apply(Implicits.scala:443)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$$anonfun$14.apply(Implicits.scala:443)
	at scala.collection.LinearSeqOptimized$class.find(LinearSeqOptimized.scala:113)
	at scala.collection.immutable.List.find(List.scala:84)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.scala$tools$nsc$typechecker$Implicits$ImplicitSearch$$typedImplicit(Implicits.scala:443)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation.rankImplicits(Implicits.scala:901)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch$ImplicitComputation.findBest(Implicits.scala:929)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:986)
	at scala.tools.nsc.typechecker.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:1375)
	at scala.tools.nsc.typechecker.Implicits$class.inferImplicit(Implicits.scala:73)
	at scala.tools.nsc.Global$$anon$1.inferImplicit(Global.scala:462)
	at scala.tools.nsc.typechecker.Implicits$class.inferImplicit(Implicits.scala:38)
	at scala.tools.nsc.Global$$anon$1.inferImplicit(Global.scala:462)
```

I'll try to put together an example in the next few days.

**Comment by mpilquist** [comment]
OK, great, I'll let you know if I think of anything in the meantime.

One option is making some of the case classes smaller -- e.g., split up `ReallyLargeCaseClass` in to multiple pieces, like `ReallyLargeCaseClass(part1: Part1, part2: Part2, part3: Part3)`.

**Comment by cridmann** [comment]
In my particular case, splitting up one of my really large case classes into smaller chunks seemed to dramatically reduce compile times.  For example, I reached a tipping point around 27 fields where adding any additional field would result in at least 30sec more of compilation time per field.

Confusingly, I have a small gist here (built off of #41) that didn't replicate this behavior in compilation performance, even though it likely has a lot more fields than my real project.  There must be some codec or something else I'm doing that isn't shown in this gist that is the culprit.  I'll continue monkeying around with it to see if I can pinpoint it.

On another interesting note, using that gist try taking out one of the int32's in the codec (so that the HList is invalid) and try to compile.  I waited 30 minutes and it didn't finish compiling (I just aborted).  I'm not sure if this is because of your library or because of shapeless.
